### PR TITLE
Add support for unix:// connections

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Adds support for HTTP over `AF_UNIX` sockets (UDS) by enabling the `unix://` scheme. Most of the functionality for this exists as part of Node's `net` module, and all this patch does is ensure the connection is created the right way.

This is useful for projects that want to use HTTP as an RPC protocol, but don't want to risk exposing a TCP socket to the public. 
If this direction seems good I'd be happy to author tests for this as well.

This patch also fixes a bug where in some cases an empty `Host` header was sent which didn't seem intentional.

Overall I hope this patch is useful. Thanks heaps!

cc/ @mcollina 

## Example

```js
const { Client } = require('undici')
const client = new Client(`unix:///tmp/hello.sock`)

client.request({
    path: '/',
    method: 'GET'
}, function (err, data) {
    if (err) throw err

    const {
        statusCode,
        headers,
        body
    } = data

    console.log('response received', statusCode)
    console.log('headers', headers)

    body.setEncoding('utf8')
    body.on('data', console.log)

    client.close()
})
```

### CODE REVIEW

**Comment by ronag** [comment]
How does a unix:// url look like? Is there a spec? I'm not familiar with it. Maybe some tests would help.

File: lib/request.js [hunk 1]
```diff
@@ -104,7 +104,7 @@ class Request extends AsyncResource {
       }
 
       if (!hostHeader) {
-        header += `host: ${hostname}\r\n`
+        header += `host: ${hostname || 'localhost'}\r\n`
```
Comment (ronag) [inline lib/request.js:107]: This should never happen? hostname should always have a value?

Comment (ronag) [reply lib/request.js:107]: I see `unix:///`

Comment (ronag) [reply lib/request.js:107]: Is `localhost` correct here?

**Comment by ronag** [comment]
https://github.com/msabramo/requests-unixsocket

**Comment by ronag** [comment]
curl choose to not support this: https://github.com/curl/curl/issues/1338

File: lib/request.js (hunk 1 above)
Comment (yoshuawuyts) [reply lib/request.js:107]: I was hitting this case -- believe if no host name is passed this value is initialized as `''`. I didn't validate we were only hitting it through the UDS path, but it seemed likely the TCP path hits this as well.

I only noticed this because my server was returning a 400 complaining `Host` hadn't been set.

**Comment by ronag** [comment]
I'm a little unsure what would make sense here in terms of url scheme... maybe we should do with an options objects like Node `{ socketPath }`.

**Comment by ronag** [comment]
@delvedor @mcollina @kmamal any thoughts?

**Comment by yoshuawuyts** [comment]
> I'm a little unsure what would make sense here in terms of url scheme...

Yeah, I wasn't too sure myself either. I picked `unix://` because it seemed like a reasonable starting point -- but something like `http+unix://` indeed makes more sense to me. It captures it's still `http`, but over a different transport than expected. One could imagine this would be enable extensions such as `http+pipe://` or `https+unix://` to work as well.

---

Regarding escaping: `http+unix:///tmp/foo.sock` is a valid WhatWG URL. [I just ran it through Firefox's URL parser](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8eb9cdfa966fd063b7e088f4665e13b5) and it seems to parse it as expected:

__code__
```rust
use url::Url;

fn main () {
    let url = Url::parse("http+unix:///tmp/foo.sock").unwrap();
    dbg!(url.scheme());
    dbg!(url.path());
    dbg!(url.domain());
}
```

__output__
```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.67s
     Running `target/debug/playground`
[src/main.rs:5] url.scheme() = "http+unix"
[src/main.rs:6] url.path() = "/tmp/foo.sock"
[src/main.rs:7] url.domain() = None
```

**Comment by ronag** [comment]
> seems to parse it as expected:

Well, `/tmp/foo.sock` isn't really a path in the http sense of path. So I'm not sure whether it makes sense. It should be the hostname, no?

**Comment by ronag** [comment]
I'm a little surprised `URL` doesn't decode `%2F`. Probably correct.

**Comment by mcollina** [comment]
I'm good with whatever scheme works.

**Comment by yoshuawuyts** [comment]
> Well, /tmp/foo.sock isn't really a path in the http sense of path. 

Yeah, there's definitely a distinction with what a url (or path or uri) is according to [RFC7230](https://tools.ietf.org/html/rfc7230#section-2.7) and according to the [WhatWG URL spec](https://url.spec.whatwg.org/). Over the years we've seen Node.js transition from the RFC7230 interpretation of URLs to the more strict but backwards-compatible WhatWG URLs, culminating in [the deprecation of the older URL APIs](https://nodejs.org/api/url.html#url_legacy_url_api).

I think the right way for us to reason about this is to follow the WhatWG URL spec. That's how both browsers and Node reason about URLs, and I think is the right way for Undici to follow as well. As I've shown in https://github.com/mcollina/undici/pull/226#issuecomment-647559681 `http+unix:///tmp/foo.sock` is parsed by WhatWG URL parsers the way we'd hope it would. And because WhatWG URLs are valid RFC7230 URLs as well, I think it's a safe choice for us to use here.

__tldr:__ The proposed URL schema is correctly parsed by a WhatWG URL parser which means we're good.

**Comment by ronag** [comment]
> is parsed by WhatWG URL parsers the way we'd hope it would

It is? Looks to me like it parses it as pathname instead of hostname? Which is not what I hoped.

i.e. if you want to make a request against a http end-point `/api/hello` the url would look somehting like:

`http+unix:///tmp/foo.sock/api/hello` which would parse everything as a pathname and we have no way to differentiate the host `/tmp/foo.sock` from the path `/api/hello`.

**Comment by yoshuawuyts** [comment]
@ronag Oh I now realize I probably misinterpreted what you said.

```rust
let url = url::Url::parse("http+unix:///tmp/foo.sock/hello").unwrap();
dbg!(url.scheme());
dbg!(url.path());
```
```text
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.63s
     Running `target/debug/playground`
[src/main.rs:5] url.scheme() = "http+unix"
[src/main.rs:6] url.path() = "/tmp/foo.sock/hello"
```

Is what you're saying that you would expect "path" to be `/hello` instead of `/tmp/foo.sock/hello`?

**Comment by ronag** [comment]
maybe we could just keep it simple and not support it through a URL scheme?, i.e.

```js
const client = new Client({
  protocol: 'unix:',
  hostname: '/tmp/foo.sock'
})

```

**Comment by yoshuawuyts** [comment]
@ronag yeah that's not a bad option. One open question tho is what we'd use to populate the `Host` header. E.g. if we do `Host: /tmp/foo.sock` then on the server side this may cause bugs. It's not uncommon for Servers to reconstruct URLs using logic like:

https://github.com/http-rs/async-h1/blob/581540a9b983d8a3fc97456d321b5bcc4eff10b7/src/server/decode.rs#L122

```rust
        Ok(Url::parse(&format!("http://{}/", host))?.join(path)?)
```

This means it'd cause the same problems on the server since URL parsing there may no longer work as expected. Should we consider escaping the hostname before we encode it as the `Host` header? Do we know of a better solution?

**Comment by ronag** [comment]
> escaping the hostname before we encode it as the Host header

I think escaping sounds like a good idea.

**Comment by ronag** [comment]
@yoshuawuyts Do you want to finish this? It's pretty close.

**Comment by yoshuawuyts** [comment]
Hey, yeah I'm planning to continue work on it this week. We need this at work, so I can spend some time on it to see it through to completion (:

**Comment by szmarczak** [comment]
What about going UNIX with TLS?

**Comment by yoshuawuyts** [comment]
> What about going UNIX with TLS?

I would suggest we scope that for a future patch. But populating the `tls` value of the options passed to [`new Client`](https://github.com/mcollina/undici#new-undiciclienturl-opts) would be the obvious approach here.

**Comment by ronag** [comment]
> I suggest we scope that for a future patch.

I think we should consider the API for that. Even if we don't implement it. How would that look? `protocol: 'unix+https:'`?

**Comment by yoshuawuyts** [comment]
@ronag I was thinking:

```js
// http+unix
const client = new Client({
  protocol: 'unix:',
  hostname: '/tmp/foo.sock'
})

// https+unix
const client = new Client({
  protocol: 'unix:',
  hostname: '/tmp/foo.sock',
  tls: tlsOpts
})
```

**Comment by yoshuawuyts** [comment]
Oh but I guess you're asking about the Host url we set? `https+unix://<escaped path>/my/sub/route` is what I was thinking.

**Comment by ronag** [comment]
@ronag I was thinking:

Hm, but that's a bit inconsistent, consider the following:

```js
const client = new Client({
  protocol: 'http:',
  hostname: '/tmp/foo.sock',
  tls: tlsOpts
})
```

Should the above error?

**Comment by yoshuawuyts** [comment]
@ronag what is the current behavior? I would assume it would error even if `hostname` wasn't set to a path.

**Comment by ronag** [comment]
> @ronag what is the current behavior? 

I think tls ignored if it's not relevant for the provided protocol.

**Comment by ronag** [comment]
@szmarczak @delvedor Do you have any good idea? This one is a little tricky.

**Comment by szmarczak** [comment]
In Got we just go UNIX if the hostname equals `unix`. E.g. `https://unix/path/to/socket` is perfectly valid.

**Comment by yoshuawuyts** [comment]
## The TLS option

> I think tls ignored if it's not relevant for the provided protocol.

That seems unexpected -- if a user went through the trouble of initializing TLS config and passing it to the constructer they would likely want to be made of the aware that the connection is in fact not secure. Forgetting to set `https` instead of `http` seems like an easy thing to get wrong.

But that's on current behavior of undici. Regarding this patch we could make it so that `protocol: "unix:"` needs to be passed in order for `hostname` to be able to be processed as a filesystem path.


## The got package

> In Got we just go UNIX if the hostname equals unix. E.g. `https://unix/path/to/socket` is perfectly valid.

In the [`got` UDS docs](https://github.com/sindresorhus/got#unix-domain-sockets) the following example is provided:

```js
const got = require('got');

got('http://unix:/var/run/docker.sock:/containers/json');

// Or without protocol (HTTP by default)
got('unix:/var/run/docker.sock:/containers/json');
```
This looks a bit differently than what was described.

**Comment by szmarczak** [comment]
> This looks a bit differently than what was described.

Can you be more precise? I cannot find a difference.

> My understanding is that this is passed through a custom parser (not WHATWG URL).

Your understanding is wrong. Got doesn't use any custom parsers.

**Comment by szmarczak** [comment]
> splits it into http:// for the protocol, unix: for the transport, /var/run/docker.sock for the host, and /containers/json for the path.

`/var/run/docker.sock` is NOT the host. It's the socket path. The `/containers/json` is a HTTP path. As you can see

https://github.com/sindresorhus/got/blob/7582d916db72f7bab2b2c162d37cfcfde08bbba7/source/core/index.ts#L1379-L1391

The `host` is empty.

Note there is NOT [yet] a spec that defines WHATWG URL schema for UNIX socket paths.

**Comment by ronag** [comment]
`'http://unix:/var/run/docker.sock:/containers/json'` is strange for me. `unix` has nothing to do with the host, it just specifies the transport protocol, while `/var/run/docker.sock` is what I would consider the host in this context, and '/containers/json' as the path.

If we can't agree on the URL version, I would suggest to go with an options version for this use case? And leave the URL parsing to higher level libraries, like got to do in whatever way they deem is best.

The following would make sense to me if we disregard URL parsing and what not:

```js
{
   protocol: 'https:',
   hostname: '/var/run/docker.sock',
   pathname: '/containers/json'
}
```

**Comment by ronag** [comment]
I wouldn't mind even removing the URL parsing from this library just to avoid these ambiguities. @mcollina

**Comment by ronag** [comment]
@yoshuawuyts Given the reaction I think we can all compromise on my solution. Would you mind updating the PR accordingly. Leave the existing URL parsing as is for now.

**Comment by mcollina** [comment]
> I wouldn't mind even removing the URL parsing from this library just to avoid these ambiguities.

That would make things simpler however push the problem down. If we could accept both an object and a string, that would be handy.

**Comment by ronag** [comment]
> That would make things simpler however push the problem down

I don't follow that.

> If we could accept both an object and a string, that would be handy.

That already works. Though for unix sockets it would only work with an object since it's unclear how to encode the socket path in a regular URL.

**Comment by ronag** [comment]
Please se https://github.com/mcollina/undici/pull/238

**Comment by mcollina** [comment]
Removing URL parsing from this library will push the problem down to our users. I'm not sure that's a good idea.

**Comment by ronag** [comment]
closing in favor of the other one.
